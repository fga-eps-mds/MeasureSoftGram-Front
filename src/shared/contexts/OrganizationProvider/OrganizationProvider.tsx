import React, { createContext, useState, useContext, ReactNode, useMemo } from 'react';

import { Organization } from '@customTypes/organization';

interface Props {
  children: ReactNode;
}

interface IOrganizationContext {
  currentOrganization: Organization;
  setCurrentOrganization: (organization: Organization) => void;
  organizationList: Organization[];
  setOrganizationList: (organizations: Organization[]) => void;
}

const OrganizationContext = createContext<IOrganizationContext | undefined>(undefined);

export function OrganizationProvider({ children }: Props) {
  const [organizationList, setOrganizationList] = useState<Organization[]>([]);
  const [currentOrganization, setCurrentOrganization] = useState<Organization>();

  const value = useMemo(
    () => ({
      currentOrganization,
      setCurrentOrganization,
      organizationList,
      setOrganizationList
    }),
    [organizationList, currentOrganization]
  );

  return <OrganizationContext.Provider value={value}>{children}</OrganizationContext.Provider>;
}

export function useOrganizationContext() {
  const context = useContext(OrganizationContext);

  if (context === undefined) {
    throw new Error('OrganizationContext must be used within a OrganizationContext');
  }

  return context;
}
