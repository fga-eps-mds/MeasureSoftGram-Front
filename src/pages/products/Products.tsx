import React, { useState } from 'react';
import Head from 'next/head';

import { Box, Container, IconButton, Menu, MenuItem, Typography } from '@mui/material';
import { MoreVert } from '@mui/icons-material';

import { NextPageWithLayout } from '@pages/_app.next';
import ConfigPage from '@pages/preConfig/ConfigPage';

import { useProductContext } from '@contexts/ProductProvider';
import { useOrganizationContext } from '@contexts/OrganizationProvider';

import getLayout from '@components/Layout';
import CardNavigation from '@components/CardNavigation';

import { Product } from '@customTypes/product';

import Skeleton from './components/Skeleton';
import { useQuery } from './hooks/useQuery';

const Products: NextPageWithLayout = () => {
  useQuery();

  const { currentOrganization } = useOrganizationContext();
  const [openConfig, setOpenConfig] = useState(false);
  const [anchorEl, setAnchorEl] = useState<null | HTMLElement>(null);
  const [selectedProduct, setSelectedProduct] = useState<Product>();

  const { productsList } = useProductContext();

  const openMenu = Boolean(anchorEl);

  const handleOpenMenu = (event: React.MouseEvent<HTMLButtonElement>) => {
    setAnchorEl(event.currentTarget);
  };

  const handleCloseMenu = () => {
    setAnchorEl(null);
  };

  const handleOpenConfig = (product: Product) => {
    setSelectedProduct(product);
    setOpenConfig(true);
    setAnchorEl(null);
  };

  const getOrganizationId = (product?: Product) => {
    if (product) {
      const url = product.organization[0];
      const urlArray = url.split('/');
      return urlArray.at(-2);
    }
    return '-1';
  };

  if (!productsList) {
    return (
      <Container>
        <Skeleton />
      </Container>
    );
  }

  return (
    <>
      <Head>
        <title> Site do MeasureSoftGram </title>
      </Head>
      <Container>
        <ConfigPage
          isOpen={openConfig}
          onClose={setOpenConfig}
          repoName={selectedProduct?.name}
          productId={selectedProduct?.id ?? '-1'}
          organizationId={getOrganizationId(selectedProduct)}
        />
        <Box display="flex" flexDirection="column">
          <Box display="flex" marginTop="40px" marginBottom="36px">
            <Typography variant="h4" marginRight="10px">
              Hi
            </Typography>
            <Typography variant="h4" fontWeight="300">
              User
            </Typography>
          </Box>

          <Box display="flex" flexWrap="wrap">
            {productsList?.map((product) => (
              <Box key={product.id} display="flex" flexDirection="row" paddingRight="20px" paddingBottom="20px">
                <CardNavigation
                  key={product.id}
                  id={product.id}
                  name={product.name}
                  url={`/products/${currentOrganization.id}-${product.id}-${product.name}`}
                />

                <IconButton color="primary" onClick={handleOpenMenu}>
                  <MoreVert />
                </IconButton>

                <Menu
                  id="basic-menu"
                  anchorEl={anchorEl}
                  open={openMenu}
                  onClose={handleCloseMenu}
                  MenuListProps={{
                    'aria-labelledby': 'basic-button'
                  }}
                >
                  <MenuItem onClick={() => handleOpenConfig(product)}>Definir pré configurações</MenuItem>
                </Menu>
              </Box>
            ))}
          </Box>
        </Box>
      </Container>
    </>
  );
};

Products.getLayout = getLayout;

export default Products;
