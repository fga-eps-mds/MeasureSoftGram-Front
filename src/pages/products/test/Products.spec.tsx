import React from 'react';
import { render } from '@testing-library/react';
import { ProductProvider } from '@contexts/ProductProvider';
import { OrganizationProvider } from '@contexts/OrganizationProvider';
import Products from '../Products';

describe('Products', () => {
  describe('Snapshot', () => {
    it('Deve corresponder ao Snapshot', () => {
      const tree = render(
        <OrganizationProvider>
          <ProductProvider>
            <Products />
          </ProductProvider>
        </OrganizationProvider>
      );
      expect(tree).toMatchSnapshot();
    });
  });
});
