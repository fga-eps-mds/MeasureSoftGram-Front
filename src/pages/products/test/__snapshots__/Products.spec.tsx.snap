// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Products Snapshot Deve corresponder ao Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiBox-root css-j7qwjs"
        >
          <div
            class="MuiBox-root css-f13xwn"
          >
            <h4
              class="MuiTypography-root MuiTypography-h4 css-obdthj-MuiTypography-root"
            >
              Hi
            </h4>
            <h4
              class="MuiTypography-root MuiTypography-h4 css-1yvpk5c-MuiTypography-root"
            >
              User
            </h4>
          </div>
          <div
            class="MuiBox-root css-2imjyh"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
    >
      <div
        class="MuiBox-root css-j7qwjs"
      >
        <div
          class="MuiBox-root css-f13xwn"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-obdthj-MuiTypography-root"
          >
            Hi
          </h4>
          <h4
            class="MuiTypography-root MuiTypography-h4 css-1yvpk5c-MuiTypography-root"
          >
            User
          </h4>
        </div>
        <div
          class="MuiBox-root css-2imjyh"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
